Parser/Parser.cs

```diff
@@ -77,16 +117,76 @@ NBlock Block ()
       Expect (BEGIN);
       List<NStmt> stmts = new ();
       while (!Match (END)) { stmts.Add (Stmt ()); Match (SEMI); }
+      Match (SEMI);
       return new (stmts.ToArray ());
    }
 
    // write-stmt =  ( "writeln" | "write" ) arglist .
    NWriteStmt WriteStmt () 
       => new (Prev.Kind == WRITELN, ArgList ());
 
+   // read-stmt = "read" "(" identlist ")" .
+   NReadStmt ReadStmt () {
+      Expect (OPEN); Expect (IDENT); Expect (CLOSE);
+      return new (Prev);
+   }
+
    // assign-stmt = IDENT ":=" expr .
    NAssignStmt AssignStmt () 
       => new (PrevPrev, Expression ());
+
+   // call-stmt =  IDENT arglist .
+   NCallStmt CallStmt ()
+      => new (Prev, ArgList ());
+
+   // if-stmt =  "if" expression "then" statement [ "else" statement ] .
+   NIfStmt IfStmt () {
+      var ifExpr = Expression ();
+      Expect (THEN);
+      List<NStmt> thenStmts = new (), elseStmts = new ();
+      if (Match (BEGIN))
+         do { thenStmts.Add (Stmt ()); Match (SEMI); } while (!Match (END));
+      else thenStmts.Add (Stmt ());
+      Match (SEMI);
+      if (Match (ELSE))
+         if (Match (BEGIN))
+            do { elseStmts.Add (Stmt ()); Match (SEMI); } while (!Match (END));
+         else elseStmts.Add (Stmt ());
+      return new NIfStmt (ifExpr, new (thenStmts.ToArray ()), new (thenStmts.ToArray ()));
+   }
+
+   // while-stmt =  "while" expression "do" statement .
+   NWhileStmt WhileStmt () {
+      var expr = Expression ();
+      Expect (DO);
+      List<NStmt> stmts = new ();
+      if (Match (BEGIN))
+         do { stmts.Add (Stmt ()); Match (SEMI); } while (!Match (END));
+      else stmts.Add (Stmt ());
+      return new (expr, new (stmts.ToArray ()));

```

Comment:
This can be: `new (exp, stmt)`
According to the language, it has only one statement.